MB-16176 - Create coverage ADR (#350)
transcom/mymove-docs: docs/adrs/0078-develop-strategy-for-ensuring-code-test-coverage.md

---
title: 0078 Develop strategy for ensuring code test coverage
description: |
  Decision outcome: The Happo Approach
---

# Develop strategy for ensuring code test coverage

**User Story:** *[MB-16176](https://dp3.atlassian.net/browse/MB-16176)* :lock:

## Problem Statement

Our current system of maintaining test coverage involves measuring the current total test coverage and ensuring that any given PR does not decrease the total coverage.

> This strategy has obvious benefits in helping to support a high test coverage, but it can also introduce unnecessary roadblocks in the PR workflow. To understand this, it's helpful to understand the scenarios that might result in a decrease in total coverage.

Reviewer: Re-wording the problem statement a bit here to better represent the issue we've faced.

```suggestion
This strategy has obvious benefits in preventing test coverage from going down, but it has also introduced roadblocks in the PR workflow. To understand this, it's helpful to understand the scenarios that might result in a decrease in total coverage.
```

### Scenarios that result in a test coverage decrease
1. **Removal of code that is _more_ tested than the total coverage.** If the total coverage is 77.0% and a block of code that is tested at 100% is deleted, the average coverage goes down.
2. **Addition of code that is _less_ tested than the total coverage and _should be tested more_.** This is the scenario that we want to address with our code coverage strategy.

> 3. **Addition of code that is _less_ tested than the total coverage and _doesn't need to be tested more_.** The difference between this scenario and scenario 2 can be subjective, but one can imagine a scenario in which a Go database model is created and doesn't require any tests.

Reviewer: Are these the only scenarios we'd like to discuss here? I think these are enough, but I'd like to hear if folks can think of another or perhaps expand on these scenarios a bit more with more examples.

Reviewer 2: **thought**: I think another example for scenario 3 would be adding logic to existing model_to_payload functions since those aren't unit tested and are only sometimes integration tested via the handler integration tests.

> Our current strategy of never allowing the total coverage to decrease attempts to address scenario 2, but it also unfairly flags scenarios 1 and 3. We also have the option to manually reset the test coverage floor, but determining whether that is appropriate (i.e. discerning which scenario is applicable to a given PR) can be difficult and time-consuming. Furthermore, any combination of the three scenarios can be applicable to a given PR -- it doesn't have to be just one -- which can make it extra difficult to decide whether a coverage decrease is legitimate.

Reviewer 3: Do we have any data for how often scenario 1 and 3 happen? In my memory, I can recall it happening twice a year which to me doesn't suggest we really have a problem. We have hundreds of PRs a year that the current process ensures we don't have a coverage decrease. Making a change for such an infrequent event doesn't really make a lot of sense to me, personally.

Reviewer 3: I'll also say that I've seen the test coverage problem come up many times, and most of the time the problem is that folks haven't rebased their PR or they need to increase test coverage. IMHO, we should just get rid of the coverage check instead of making it optional, as I believe folks will just ignore it.

Reviewer: @ahobson Thanks Drew. Is this supposed to read "**should** just get rid of" _or_ "**shouldn't** just get rid of"?

Reviewer 3: @rogeruiz IMHO, if we think the test coverage problem is major enough to make the checks optional, we should just remove the checks entirely.

Reviewer 3: I've been thinking about this, and I think I'm wrong. I don't have any data to support my position that making the check optional will actually decrease test coverage.

This proposed change is one that we could easily undo if we needed to.

How about this: We note the coverage before putting this change in, and then one month later, at BE/FE check in, we report on the coverage numbers and how they compare to before the change. We can this discuss in the check in if the changes are concerning or not and if we want to keep the checks optional. It can become something we bring up monthly at the check in to ensure we aren't drifting the wrong way.

Thoughts?

Reviewer 3: I updated the PR to record the coverage on merge to main to the govdev environment and we now have a dashboard.

<img width="611" alt="image" src="https://github.com/transcom/mymove-docs/assets/21983/586ccd7f-6a9e-4526-ab28-0c402be26fa4">

Author: @ahobson I think this plan makes a lot of sense -- I can mention in the chosen option that we will review it after a month. I should also note that a fairer comparison would be with the coverage if we didn't make the check optional (which we can't know), which would also potentially decrease every time we manually reset it.

Reviewer 2: **thought**: Good point @michaelvaldes-truss. I wonder if to make that comparison match better we could compare the new change in test coverage against historical data of how coverage has changed previously with the required check in place since we have reset the cache a few times.

Author: I also want to address the very valid concerns expressed in comments on this PR -- that making the check optional would make it too easy for coverage to decrease. My assessment is that the current strategy already allows for decreases when folks manually reset the limit, and this can happen for very valid reasons (scenarios 2 and 3 described in the ADR). Analyzing coverage decreases is difficult to do automatically, so the intention of the Happo Approach was to put it entirely in the hands of the PR reviewers and also make it easier to access information relevant to that analysis.

I added these points to the ADR and also introduced a third option, Lenience for Tiny Coverage Decreases, which is a middle ground that attempts to partially address these concerns. 

@YanZ777 @kctruss

Reviewer 4: I do want to note that while I support the third option presented - there's a small chance there could be a .2% difference if you pull main to fix a .1% diff in coverage. That's what happened to me this last go round with coverage issues. [Little more detail here.](https://ustcdp3.slack.com/archives/C04U4TWDLKG/p1686263061819939?thread_ts=1686244694.130929&cid=C04U4TWDLKG)

## Test Coverage Objectives
To improve upon the current coverage strategy and address the scenarios above, we propose that our coverage strategy aims to address the following objectives:
- Reduce the amount of new and updated code that is not tested enough (scenario 2)
- Provide folks with the flexibility to easily override strict coverage enforcement in scenarios that don't require it
- Provide folks with easier access to context that allows them to distinguish between less-tested code that should be tested more (scenario 2) and less-tested code that doesn't need to be tested more (scenario 3)
- Address the above objectives with little effort

## Considered Options

* *Do nothing*
* *The Happo Approach*

## Decision Outcome

<!-- * Chosen Option:  -->

> Chosen Option: *The Happo Approach*

Reviewer 5: **question:** Will this ADR be reassessed in the future? Is the plan to have the test coverage be optional forever?

Reviewer 2: It seems like maybe we'll re-assess after a month to see if things are going well: https://github.com/transcom/mymove-docs/pull/350/files#r1221538465

Though if things aren't going how we'd like them to, we'd likely then have to decide how we want to correct things.

Author: Thanks -- I added a note in the ADR to reflect this!

Our approach to Happo changes is currently the following: the Happo PR check is not a required check, but it reports failures and includes an easily-accessible link on the PR to Happo diffs which provide context around visual changes. Something similar could be implemented for code coverage checks with the following steps:
1. Make the code coverage checks optional so that failures do not block a PR

> 2. Generate a link automatically on the PR (perhaps through a comment) that provides more context behind the failure (including, but not limited to, the go-coverage.html artifact generated in circle-ci)

Reviewer: This link should also point to documentation around the coverage tests and how to increase it. I think it's also possible to include this as a third option that may be the best of keeping the tests and having a GH Action that alerts authors to read the specific documentation about how to interpret the code coverage results for the client and the server and also reset the cache in a way that can be tracked / discussed when it comes up.

Reviewer 2: **thought**: +1 to the link to docs and that this could be its own third option, even if we are already implementing the happo option.

Author: That makes sense to me -- I'll gauge folk's feelings at the FE/BE check-in and update the ADR accordingly.

* `+` *Low-medium effort*
* `+` *Provides a way to maintain high test coverage*
* `+` *Does not unnecessarily block PRs*
* `+` *Provides easily-accessible context behind coverage decreases, at least with server coverage*
* `-` *Requires a cultural enforcement of coverage maintenance*

## Pros and Cons of the Alternatives

### *Do nothing*

Continue blocking PRs that decrease test coverage.

* `+` *No implementation effort*

> * `+` *Provides a way to waintain high test coverage with stricter enforcement*

Reviewer 6: **typo:** `waintain` => `maintain`

Author: Thanks for catching this -- we've got to waintain our spelling standards too.

> * `+` *Provides a way to maintain high test coverage with stricter enforcement*

Reviewer: ```suggestion
* `+` *Provides a way to maintain current test coverage and never decrease it*
```
